src/storage/mod.rs

```diff
@@ -191,6 +191,108 @@ where
     }
 }
 
+/// An entry to a storage which has a component associated to the entity.
+pub struct OccupiedEntry<'a, 'b: 'a, T: 'a, D: 'a> {
+    entity: Entity,
+    storage: &'a mut Storage<'b, T, D>,
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Get a reference to the component associated with the entity.
+    pub fn get(&self) -> Option<&T> {
+        self.storage.get(self.entity)
+    }
+}
+
+impl<'a, 'b, T, D> OccupiedEntry<'a, 'b, T, D>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Get a mutable reference to the component associated with the entity.
+    pub fn get_mut(self) -> &'a mut T {
+        match self.storage.get_mut(self.entity) {
+            Some(component) => component,
+            None => { panic!("`OccupiedEntry` on an entity without a component."); },
+        }
+    }
+
+    /// Inserts a value into the storage and returns the old one.
+    pub fn insert(&mut self, component: T) -> T {
+        match self.storage.insert(self.entity, component) {
+            InsertResult::Updated(old) => old,
+            InsertResult::Inserted => { panic!("`OccupiedEntry` on an entity without a component."); },
+            InsertResult::EntityIsDead(_) => { panic!("`OccupiedEntry` on a dead entity: {:?}.", self.entity); },
+        }
+    }
+
+    /// Removes the component from the storage and returns it.
+    pub fn remove(self) -> T {
+        match self.storage.remove(self.entity) {
+            Some(old) => old,
+            None => { panic!("`OccupiedEntry` on an entity without a component."); },
```

Comment:
Meh, the whole sense of Entry API is that you usually don't do any checks twice which is the case for all these methods :(